sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2808,6 +2808,53 @@ class Dataset[T] private[sql](
     }
   }
 
+  /**
+   * (Scala-specific)
+   * Returns a new Dataset with a columns renamed.
+   * This is a no-op if schema doesn't contain existingName.
+   *
+   * `colsMap` is a map of existing column name and new column name.
+   *
+   * @group untypedrel
+   * @since 3.4.0
+   */
+  def withColumnsRenamed(colsMap: Map[String, String]): DataFrame = {
+    val resolver = sparkSession.sessionState.analyzer.resolver
+    val output = queryExecution.analyzed.output
+
```

Comment:
a simpler approach
```
val projectList = output.map { attr =>
  if (resolver...) Alias(attr, ...) else attr
}
withPlan(Project(projectList, logicalPlan))
```

Reply:
I will give it a try, thanks for the review!

Reply:
New implementation is in. @cloud-fan is that what you had in mind?